### PR TITLE
add github ci compile check for ascent

### PR DESCRIPTION
**Pull request by cyrush**
Adds github ci check that complies WarpX with Ascent=ON, using an Ascent container. 

From here we could run a small test, render, image and provide that image as an artifact.

### CODE REVIEW

File: .github/workflows/ascent.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+name: 🐧 In Situ Vis with Ascent
+
+on: [push, pull_request]
+
+concurrency:
+  group: ${{ github.ref }}-${{ github.head_ref }}-insituvis
```
Comment (ax3l) [inline .github/workflows/ascent.yml]: Either we make this a unique one
```suggestion
  group: ${{ github.ref }}-${{ github.head_ref }}-ascent
```

or even better - we put them into one `.yml` file (adding the new `jobs` key alongside the existing one :) )

With the current `concurrency` `group` name, both the `ascent.yml` and the `insitu.yml` would cancel each other.

File: .github/workflows/insitu.yml [hunk 2]
```diff
@@ -30,3 +30,23 @@ jobs:
...
+      CXX: g++
+      CC: gcc
+      CMAKE_PREFIX_PATH: /home/user/ascent/install-debug/lib/cmake/
+    container:
+      image: alpinedav/ascent:ascent-ubuntu-18_2021-10-23-shaba7872
+    steps:
+    - uses: actions/checkout@v2
+    - name: Configure
+      run: |
+        . ../s_env.sh && cmake -S . -B build               \
```
Comment (ax3l) [inline .github/workflows/insitu.yml]: Purely aesthetic to clean up later on: you can set up the environment (`s_env.sh`) and the `CMAKE_PREFIX_PATH` hint in your docker image similar to what I did here, by adding a script in `/etc/profile.d/` and using an entry point that is a login shell (if needed):
- https://github.com/ComputationalRadiationPhysics/picongpu/blob/f73109281b234b1d0c8eb646e44d680d5f7ce95d/share/picongpu/dockerfiles/ubuntu-2004/Dockerfile#L62-L79
- https://github.com/ComputationalRadiationPhysics/picongpu/blob/f73109281b234b1d0c8eb646e44d680d5f7ce95d/share/picongpu/dockerfiles/ubuntu-2004/Dockerfile#L122-L123

File: .github/workflows/insitu.yml [hunk 3]
```diff
@@ -30,3 +30,23 @@ jobs:
     - name: Build
       run: |
         cmake --build build -j 2
+  ascent:
```
Comment (ax3l) [inline .github/workflows/insitu.yml:34]: Just some newline / spacing for better readability:
```suggestion

  ascent:
```

File: .github/workflows/insitu.yml [hunk 4]
```diff
@@ -30,3 +30,23 @@ jobs:
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: Configure
+      run: |
+        . ../s_env.sh && cmake -S . -B build               \
+                                    -DWarpX_ASCENT=ON      \
+                                    -DWarpX_COMPUTE=NOACC
+    - name: Build
+      run: |
+        . ../s_env.sh && cmake --build build -j 2
```
Comment (ax3l) [inline .github/workflows/insitu.yml]: Just an end-of-file newline:
```suggestion
        . ../s_env.sh && cmake --build build -j 2
```

File: .github/workflows/insitu.yml (hunk 2 above)
Comment (cyrush) [reply .github/workflows/insitu.yml]: yes, thanks for sharing the magic! I will update the docker image during my next build to to simplify the process

File: .github/workflows/insitu.yml (hunk 4 above)
Comment (cyrush) [reply .github/workflows/insitu.yml]: I can't commit this one for some reason.

My first pass I had an end-of-file nl, but i failed a style check so I removed it :-)

Comment (ax3l) [reply .github/workflows/insitu.yml]: You can just not add empty spaces at the end of a line in our style ^^
Empty newlines are ok :)

**Comment by cyrush** [comment]
@ax3l i have the container build test for Ascent sorted out. 

Do you suggest we also run a small test and render a pictured as part of the process, if so what test would you recommend?

**Comment by cyrush** [comment]
- [x] Cyrus: Add newline to end of CI yaml file.

**Comment by ax3l** [comment]
@cyrush awesome, thanks a lot!

We could use the 3D laser-wakefield (LWFA) test: https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Physics_applications/laser_acceleration/inputs_3d and render out in a single scene:
- iso-contour of Ey (laser pulse)
- volume rendering of Ey (just to check it works)
- particle species: "electrons" colored by pz?

**Comment by cyrush** [comment]
@ax3l sounds like a plan, I will tackle that soon.

**Comment by cyrush** [comment]
I have the tests running :-)

I put the input deck and ascent actions file in `.github/workflows/resources`, not sure if this is a good place so please let me know if I should move elsewhere. 

The setup also provides artifacts (pngs and blueprint files), you can see an example here:
https://github.com/ECP-WarpX/WarpX/actions/runs/1419129561

**Comment by ax3l** [comment]
Thanks a lot, that's fantastic!

> I put the input deck and ascent actions file in .github/workflows/resources, not sure if this is a good place so please let me know if I should move elsewhere.

Nice, I added some house-keeping updates in https://github.com/Alpine-DAV/WarpX/pull/1

> The setup also provides artifacts

Awesome! For the Ey volume rendering, can you please set the color bar to be
- symmetric around zero and
- transparent for zero

? :)

**Comment by ax3l** [comment]
Ups, let's rename `Examples/Physics_applications/laser_acceleration/inputs_3d_ascent_test_ascent_actions.yaml` to `Examples/Physics_applications/laser_acceleration/3d_ascent_actions.yaml` or so :)

File: .github/workflows/insitu.yml [hunk 5]
```diff
@@ -30,3 +30,43 @@ jobs:
...
+            -DWarpX_ASCENT=ON \
+            -DWarpX_COMPUTE=NOACC
+    - name: Build
+      run: |
+        . /ascent_docker_setup_env.sh
+        cmake --build build -j 2
+    - name: Test
+      run: |
+        cp Examples/Physics_applications/laser_acceleration/inputs_3d .
+        cp Examples/Physics_applications/laser_acceleration/inputs_3d_ascent_test_ascent_actions.yaml ascent_actions.yaml
```
Comment (ax3l) [inline .github/workflows/insitu.yml]: After renaming the file, this line needs to read:
```suggestion
        cp Examples/Physics_applications/laser_acceleration/3d_ascent_actions.yaml
```

The background is that we check in CI that all WarpX `inputs_...` files are covered by CI tests :)

File: .github/workflows/insitu.yml [hunk 6]
```diff
@@ -30,3 +30,43 @@ jobs:
...
+
+        mpiexec -n 2 ./build/bin/warpx \
+          inputs_3d_ascent_test        \
```
Comment (ax3l) [inline .github/workflows/insitu.yml]: Ups, forgot to rename:
```suggestion
          inputs_3d                    \
```

File: .github/workflows/insitu.yml (hunk 5 above)
Comment (cyrush) [reply .github/workflows/insitu.yml]: Yes, that bit me before (that's why I added the smile above:)

**Comment by cyrush** [comment]
@ax3l  I updated the color table for the volume rendering example, here is the output (from the actions artifact)


ey_volrend_140

Let me know if there are other tweaks we should make.

**Comment by ax3l** [comment]
@cyrush thanks a lot, awesome! Can we rotate the volume rendering figure by 90 degrees in the horizontal axis please? :)
(Positive z values to the right, if possible :) )

**Comment by cyrush** [comment]
ey_volrend_40
uz_pc_40
ey_contour_pc_40

@ax3l  How do these look?

Note: For contour, the extents are different post contour - so that is why it looks different from the other two.

**Comment by ax3l** [comment]
Awesome! :star_struck: 

Thanks a lot! :rocket: :sparkles:

**Review by ax3l — APPROVED** [review]

**Pull request merged**